JDK-8318761: MessageFormat pattern support for CompactNumberFormat, ListFormat, and DateTimeFormatter

justin-curtis-lu opened the pull request:
Please review this change which updates MessageFormat to support String patterns for the following subformats: ListFormat, CompactNumberFormat, and DateTimeFormatter. See the [CSR](https://bugs.openjdk.org/browse/JDK-8319344) for further detail.

In addition, the MessageFormat implementation for `applyPattern()` was changed from the int enum pattern to support two new enums: _FormatType_, _FormatStyle_

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires CSR request [JDK-8319344](https://bugs.openjdk.org/browse/JDK-8319344) to be approved

### Issues
 * [JDK-8318761](https://bugs.openjdk.org/browse/JDK-8318761): MessageFormat pattern support for CompactNumberFormat, ListFormat, and DateTimeFormatter (**Enhancement** - P4)
 * [JDK-8319344](https://bugs.openjdk.org/browse/JDK-8319344): MessageFormat pattern support for CompactNumberFormat, ListFormat, and DateTimeFormatter (**CSR**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16588/head:pull/16588` \
`$ git checkout pull/16588`

Update a local copy of the PR: \
`$ git checkout pull/16588` \
`$ git pull https://git.openjdk.org/jdk.git pull/16588/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16588`

View PR using the GUI difftool: \
`$ git pr show -t 16588`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16588.diff">https://git.openjdk.org/jdk/pull/16588.diff</a>

</details>

jmehrens commented:
@justin-curtis-lu I noticed that java.text.DateFormat instance allows a [format target of java.util.Date or java.lang.Number](https://github.com/openjdk/jdk/blob/891d8cfaf2fc0636bfe8f864cd010fb71266d723/src/java.base/share/classes/java/text/DateFormat.java#L342) but [java.time.format.ClassicFormatter](https://github.com/openjdk/jdk/blob/891d8cfaf2fc0636bfe8f864cd010fb71266d723/src/java.base/share/classes/java/time/format/DateTimeFormatter.java#L2267) only allows java.time.temporal.TemporalAccessor.  This would imply that if simple date format patterns of message format are updated to use the new temporal formats the code could fail at runtime as the format target is more strict for temporal formats.  The java.util.Date and subclasses from the sql package would also not work with the temporal formats even though they have conversion methods to construct a TemporalAccessor.

The [org.eclipse.angus.mail.util.logging.CollectorFormatter](https://eclipse-ee4j.github.io/angus-mail/docs/api/org.eclipse.angus.mail/org/eclipse/angus/mail/util/logging/CollectorFormatter.html#format(java.util.logging.LogRecord)) internally uses the MessageFormat with long values (java.lang.Number) to represent minMillis, maxMillis, startTime, and currentTime which is passed as is to the format method of the MessageFormat.  Since this is a logging formatter the main functionality comes from defining a message format pattern.

What are your thoughts on allowing Number and conversions to TemporalAccessor for legacy data types with the new patterns? 
If those data types are forbidden for new patterns then, I don't think it is a blocker for me with the CollectorFormatter as I should be able to come up with a patch to allow both new and old patterns.

justin-curtis-lu commented:
@jmehrens, thanks for raising your concerns. 

If I understand correctly, you’re worried that the existing ‘_date_’ and ’_time_’ patterns will replace the DateFormat under the hood with a DateTimeFormatter (ClassicFormat). That isn’t the case, the intention of this patch is to introduce additional patterns ‘_j_date_’ and ‘_j_time_’ that support DateTimeFormatter (ClassicFormat) in addition to ‘date’ and ’time’ still supporting DateFormat. Causing a compatibility issue is not desired at all, we want users to be able to format date/time whether it be from java.util.Date or java.time.

With this patch, if you have a MessageFormat with a ‘date’ pattern it will still use a DateFormat, and thus you can still format a `Number` with that pattern. Only if you were to change 'date' to 'j_date' would MessageFormat use a ClassicFormat instead, which would then only format `TemporalAccessor`.

Please see [JDK-8319344](https://bugs.openjdk.org/browse/JDK-8319344),  which is more up to date with the current intentions and provides a greater understanding of the purpose of this patch.

Please let me know if I interpreted/answered your question correctly.

jmehrens commented:
@justin-curtis-lu Thanks for the clarifications and the link to the CSR.  From what I understand from the discussion so far is that it is intentional that Number and java.util.Date are not supported under the new j_date and j_time.

Maybe add a 4th example to "Add a section to clarify formatting java.time objects versus java.util.Date"
````
Object[] arg = {System.currentTimeMillis()};
var fmt = new MessageFormat("The date was {0,date,full}");
fmt.format(arg); // returns "The date was Thursday, November 16, 2023"
````
I could at least point to that when fielding questions on CollectorFormatter.

Why not have the MessageFormatter support the older data types with the newer patterns?  The java.util.Formatter is able to handle number, date, and ZonedDateTime:
````
System.out.printf("%1$tc%n", System.currentTimeMillis());
System.out.printf("%1$tc%n", new java.util.Date());
System.out.printf("%1$tc%n", ZonedDateTime.now());
System.out.printf("%1$tc%n", ZonedDateTime.ofInstant(Instant.now(), ZoneId.systemDefault()));
====
Wed Nov 29 22:02:08 CST 2023
Wed Nov 29 22:02:08 CST 2023
Wed Nov 29 22:02:08 CST 2023
Wed Nov 29 22:02:08 CST 2023
````

The java.util.Date and subclasses have built in methods that perform conversions to TemporalAccessor.  Number could be treated as `ZonedDateTime.ofInstant(Instance.ofEpochMilli(number), ZoneId.systemDefault())`

The reason I was leaning towards this approach is because of the example class [CollectorFormatter](https://eclipse-ee4j.github.io/angus-mail/docs/api/org.eclipse.angus.mail/org/eclipse/angus/mail/util/logging/CollectorFormatter.html#format(java.util.logging.LogRecord)).  Such a class only allows the API user to control the pattern of the message format and not the argument types.  So it is not that the change is incompatible is it just that there is a new way to break the CollectorFormatter that I'll have to decide how to handle.  Ideally I would like to support both formats.  The CollectorFormatter wouldn't break if MessageFormatter did the conversions to the arguments.  Since the MessageFormatter is parsing the pattern it should have knowledge if it one of the indexed formatters is a ClassicFormatter and the matching argument is not TemporalAccessor to perform a conversion before passing the argument to the downstream formatter.

justin-curtis-lu commented:
@jmehrens Thanks for responding, I think I understand your perspective. Since the user has no control over the arguments they cant make the conversions. As the majority of the arguments are `Number`, the new patterns will break your application if a user uses them, which you need to figure out how you want to handle.

`MessageFormat` is designed to simply hold subformats and it is up to the subformats to implement their own logic on how to handle their arguments. The specification has a table which indicates which `Format` subclass will be returned from the associated pattern. If we then add custom logic where certain `Format` subclasses make conversions on the argument type, it adds more complexity to the class and is no longer consistent with the table. We think it is also more clear, with date and time reserved for java.util.Date, and _j_date_ and _j_time_ reserved for java.time. We would prefer to keep them separate.

It would make the most sense for the user to make the conversion from `Number` to `TemporalAccessor` if they want to use _j_date_, not the MessageFormat itself (although I understand this is not possible for your case, as the user does not control the argument to format).

I assume for _head_ and _tail_ properties in `getTail()`, you simply disallow all subformat patterns for those indexes as the arguments are strings. Could you implement similar logic to disallow these new subformat patterns for all the indexes? Or implement the conversion from Number to TemporalAccessor yourself, if a _j_date/time_ pattern is supplied. Also, wouldn't the `ListFormat` patterns have the same issue for you? `ListFormat` will never be able to format `Number`, as it expects a List or Array, and there is no sensible conversion. I assume `CompactNumberFormat` is fine.

I'd like to cause the least trouble for you, but this class already supports any JDK Format subclass through the MessageFormat setter methods, and the pattern syntax needs to catch up.

justin-curtis-lu commented:
@jmehrens Likewise, thanks for bringing up your concerns.

> So it sounds like if I wanted to pursue this further it would be an RFE against ClassicFormatter as everything I'm suggesting is conversion work that if accepted would belong in that class. The end result of updating ClassicFormatter would then make MessageFormat align with java.util.Formatter with regards to various old and new data types.

Yes I would believe so. However, I believe when j.time was introduced, there was an intentional decision to not support j.u.Date and j.u.Calendar in this manner. Someone with more familiarity of java.time would take a look at the RFE if one is filed and make that decision.

Another possibility would have been extending ClassicFormat so that you could override the format methods with your own implicit conversions. But unfortunately ClassicFormat is not a public class, and MessageFormat does not support custom user defined patterns (although I believe there is an RFE for this) for say an extended ClassicFormat child class.

RogerRiggs commented on src/java.base/share/classes/java/text/MessageFormat.java:
```diff
@@ -1651,6 +1779,88 @@ private static void copyAndFixQuotes(String source, int start, int end,
+        LIST("list"),
+
+        // Pre-defined DateTimeFormatter types
+        BASIC_ISO_DATE("basic_iso_date"),
+        ISO_LOCAL_DATE("iso_local_date"),
+        ISO_OFFSET_DATE ("iso_offset_date"),
+        ISO_DATE("iso_date"),
+        ISO_LOCAL_TIME("iso_local_time"),
+        ISO_OFFSET_TIME("iso_offset_time"),
+        ISO_TIME("iso_time"),
+        ISO_LOCAL_DATE_TIME("iso_local_date_time"),
+        ISO_OFFSET_DATE_TIME("iso_offset_date_time"),
+        ISO_ZONED_DATE_TIME("iso_zoned_date_time"),
+        ISO_DATE_TIME("iso_date_time"),
+        ISO_ORDINAL_DATE("iso_ordinal_date"),
+        ISO_WEEK_DATE("iso_week_date"),
+        ISO_INSTANT("iso_instant"),
+        RFC_1123_DATE_TIME("rfc_1123_date_time");
```
Don't map these to lower case, keep the case exactly as declared in DateTimeFormatter.
It will be clearer that they are pre-defined constants.

justin-curtis-lu commented:
Hi @RogerRiggs  and @naotoj ,

There has been discussion on what the preferred `FormatType` might be for the new _java.time_ types.

The names should be well-defined, but at the same time, not too long, as these patterns are typed out by the user, and the intention is to serve as quick way to achieve the desired MessageFormat.

Previously, the prefix was _j_time_ which may not be well-defined for an unfamiliar user. Currently the draft PR has them defined as _java_temporal_time_, while being well-defined, this is probably too long for a short hand pattern. 

_java_time_ may be a happy medium, but _java_time_time_ sounds odd.

How do you both feel about `FormatType` prefix names such as _jt_time_ (where jt stands for java.time) or _temporal_time_? I am also open to other thoughts and suggestions; would like to know what you both think.

RogerRiggs commented:
> There has been discussion on what the preferred `FormatType` might be for the new _java.time_ types.

The DateTimeFormatter formats might be easily recognized and short as "dtf_date", "dtf_time", and "dtf_datetime".
The table includes the full name and invocation and is easy to see the mapping.

I'd also be inclined to add all of the predefined DateTimeFormatter formats to the table, instead of in a separate paragraph.

naotoj commented:
I am fine with either `jt_date/time/datetime` or `dtf_date/time/datetime` if they refer to `DateTimeFormatter` explicitly. As to the predefined ones, I think it would be too busy to include all of them in the table. My $0.02

justin-curtis-lu commented:
Hi @RogerRiggs,

I updated the table to include the pre-defined DateTimeFormatters as: `<td>The {@code pre-defined DateTimeFormatter(s)} are used 1-1 as a {@code FormatType} | {@link DateTimeFormatter#BASIC_ISO_DATE BASIC_ISO_DATE}`... 
I added the "1-1" wording for clarification, since the separate paragraph was removed.

I also updated the FormatType to `dtf_time/date/datetime` as we discussed.

RogerRiggs commented:
Looks ok, I would leave out 1-1 and just say "are used as a FormatType".

justin-curtis-lu commented:
Thanks @RogerRiggs, removed in https://github.com/openjdk/jdk/pull/16588/commits/1d8ba18f427f33b3f4f8694ea08c664ecbc4f6ae

justin-curtis-lu commented:
Closed in place of #17663
